### ISSUE TITLE
Convolution of masked arrays doesn't work

### ISSUE THREAD

**Issue by julietbravo** [numpy/numpy#6458]
For example:

```
a = np.random.random(10)
a[5] = 1e9
a = np.ma.masked_equal(a, 1e9) 
b = np.convolve(a, np.ones(4)/4.)
```

gives incorrect results.

**Comment by jjhelmus** [comment]
Labeling as Enhancement as convolution of masked array would require the addition of a `np.ma.convolve` function

**Comment by eric-wieser** [comment]
There seem to be multiple sensible implementation choices here:

Consider:

```python
a = np.ma.masked_equal(np.arange(5), 2)
b = np.ones(2)
```

* Masked values are contagious, `np.convolve(a, b) == [0, 1, --, --, 7, 4]` (equivalent to `np.convolve(a.filled(np.nan), b)`)
* Masked values are silent, `np.convolve(a, b) == [0, 1, 2, 4, 7, 4]`

Which of these fits better with the intent of a masked array?

**Comment by eric-wieser** [comment]
Here's a very basic implementation that assumes that both arguments are masked arrays, and implements both the behaviours described above

```python
def convolve(a, b, contagious=True):
    data = np.convolve(a.data, b.data)
    if not contagious:
        # results which are not contributed to by any pair of valid elements
        mask = ~np.convolve(~a.mask, ~b.mask) 
    else:
        # results which are contributed to by either item in any pair being invalid
        mask = np.convolve(a.mask, np.ones(b.shape, dtype=np.bool)
        mask |= np.convolve(np.ones(a.shape, dtype=np.bool), b.mask)
    return np.masked_array(data, mask)
```

**Comment by eric-wieser** [comment]
@jjhelmus: This can be closed thanks to #7922

**Comment by jjhelmus** [comment]
Thanks for the update @eric-wieser.  Closing.
